lib/mongo/server/connectable.rb

```diff
@@ -88,10 +88,11 @@ def ensure_connected
         ensure_same_process!
         connect!
         begin
-          yield socket
-        rescue Exception => e
-          disconnect!
-          raise e
+          res = yield socket
+          success = true
+          res
+        ensure
+          success or disconnect!
```

Comment:
I see what you're saying. That's a good point so using the success variable is safer.